Add tests for CUDAFuture

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56518 Add tests for CUDAFuture**
* #56516 Support custom Python classes in CUDAFuture

I don't think we have any tests for CUDAFuture (I couldn't find any, and I didn't write any in the past). I think especially for the two latest features added by this stack we should have a test to ensure they properly work and to catch regressions. (These tests also add indirect coverage for the more "basic" features of CUDAFuture).

I didn't know how/where to add tests for C++ ATen stuff, so instead I added these tests to the Python RPC suite, using the torch.futures.Future wrapper. (It made sense in my mind because RPC is the main user of CUDAFuture). I'll gladly accept pointers to better ways of doing this.

Differential Revision: [D27887191](https://our.internmc.facebook.com/intern/diff/D27887191/)

lw committed:
Add tests for CUDAFuture

I don't think we have any tests for CUDAFuture (I couldn't find any, and I didn't write any in the past). I think especially for the two latest features added by this stack we should have a test to ensure they properly work and to catch regressions. (These tests also add indirect coverage for the more "basic" features of CUDAFuture).

I didn't know how/where to add tests for C++ ATen stuff, so instead I added these tests to the Python RPC suite, using the torch.futures.Future wrapper. (It made sense in my mind because RPC is the main user of CUDAFuture). I'll gladly accept pointers to better ways of doing this.

Differential Revision: [D27887191](https://our.internmc.facebook.com/intern/diff/D27887191/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -5635,3 +5660,81 @@ def test_rref_to_here_synchronization_cross_device(self):
                 self.assertEqual(actual, expected)
 
         rpc.shutdown()
+
```
Shall we also add a test for the following error message?

```c++
    TORCH_CHECK_VALUE(
        excessDevices.empty(),
        "The result contained tensors residing on devices ",
        formatSetOfDevices(excessDevices),
        " which are not among the expected devices ",
        formatSetOfDevices(*devices_));
```

lw replied:
Done, and in fact that caught a bug (i.e., `DeviceIndex`es are `int8_t`s, a.k.a., chars, hence outputting them to a stringstream won't format them as numbers).

lw committed:
Update on "Add tests for CUDAFuture"


I don't think we have any tests for CUDAFuture (I couldn't find any, and I didn't write any in the past). I think especially for the two latest features added by this stack we should have a test to ensure they properly work and to catch regressions. (These tests also add indirect coverage for the more "basic" features of CUDAFuture).

I didn't know how/where to add tests for C++ ATen stuff, so instead I added these tests to the Python RPC suite, using the torch.futures.Future wrapper. (It made sense in my mind because RPC is the main user of CUDAFuture). I'll gladly accept pointers to better ways of doing this.

Differential Revision: [D27887191](https://our.internmc.facebook.com/intern/diff/D27887191/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -501,6 +502,30 @@ def MyConvNetForMNIST(device):
 
+# A custom Python class that contains a tensor, needed to see if we correctly
+# use the Python pickler to extract tensors from non-IValue-convertible types.
+class TensorWrapper:
```
Do we plan to address or silent this lint error?

mrshenli reviewed (approved):
LGTM!

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -5635,3 +5660,111 @@ def test_rref_to_here_synchronization_cross_device(self):
+                future.set_result(wrapper(tensor))
+            with torch.cuda.stream(another_stream):
+                self.assertTrue(torch.equal(
+                    unwrapper(future.wait()).cpu(), torch.ones((100,))
```
Not an expert on cuda stream synchronization so this comment might not be relevant. But should we be doing this check on the device instead of moving to CPU? AFAIK, moving to CPU triggers a device synchronize so all streams are synchronized when `.cpu()` is called. 

What we want to test is that `event.block()` in postWaitHook is blocking the current stream appropriately right?

lw replied:
Good point. I am not sure if a `.cpu()` call causes a "full" device syncs (all kernels on all streams), or just a sync for the current stream. Though we shouldn't take any chances in tests, hence I changed this to do the comparison on GPU and only copy the final boolean result back to CPU.

lw committed:
Update on "Add tests for CUDAFuture"


I don't think we have any tests for CUDAFuture (I couldn't find any, and I didn't write any in the past). I think especially for the two latest features added by this stack we should have a test to ensure they properly work and to catch regressions. (These tests also add indirect coverage for the more "basic" features of CUDAFuture).

I didn't know how/where to add tests for C++ ATen stuff, so instead I added these tests to the Python RPC suite, using the torch.futures.Future wrapper. (It made sense in my mind because RPC is the main user of CUDAFuture). I'll gladly accept pointers to better ways of doing this.

Differential Revision: [D27887191](https://our.internmc.facebook.com/intern/diff/D27887191/)

[ghstack-poisoned]

lw committed:
Update on "Add tests for CUDAFuture"


I don't think we have any tests for CUDAFuture (I couldn't find any, and I didn't write any in the past). I think especially for the two latest features added by this stack we should have a test to ensure they properly work and to catch regressions. (These tests also add indirect coverage for the more "basic" features of CUDAFuture).

I didn't know how/where to add tests for C++ ATen stuff, so instead I added these tests to the Python RPC suite, using the torch.futures.Future wrapper. (It made sense in my mind because RPC is the main user of CUDAFuture). I'll gladly accept pointers to better ways of doing this.

Differential Revision: [D27887191](https://our.internmc.facebook.com/intern/diff/D27887191/)

[ghstack-poisoned]

lw committed:
Update on "Add tests for CUDAFuture"


I don't think we have any tests for CUDAFuture (I couldn't find any, and I didn't write any in the past). I think especially for the two latest features added by this stack we should have a test to ensure they properly work and to catch regressions. (These tests also add indirect coverage for the more "basic" features of CUDAFuture).

I didn't know how/where to add tests for C++ ATen stuff, so instead I added these tests to the Python RPC suite, using the torch.futures.Future wrapper. (It made sense in my mind because RPC is the main user of CUDAFuture). I'll gladly accept pointers to better ways of doing this.

Differential Revision: [D27887191](https://our.internmc.facebook.com/intern/diff/D27887191/)

[ghstack-poisoned]

lw committed:
Update on "Add tests for CUDAFuture"


I don't think we have any tests for CUDAFuture (I couldn't find any, and I didn't write any in the past). I think especially for the two latest features added by this stack we should have a test to ensure they properly work and to catch regressions. (These tests also add indirect coverage for the more "basic" features of CUDAFuture).

I didn't know how/where to add tests for C++ ATen stuff, so instead I added these tests to the Python RPC suite, using the torch.futures.Future wrapper. (It made sense in my mind because RPC is the main user of CUDAFuture). I'll gladly accept pointers to better ways of doing this.

Differential Revision: [D27887191](https://our.internmc.facebook.com/intern/diff/D27887191/)

[ghstack-poisoned]

lw committed:
Update on "Add tests for CUDAFuture"


I don't think we have any tests for CUDAFuture (I couldn't find any, and I didn't write any in the past). I think especially for the two latest features added by this stack we should have a test to ensure they properly work and to catch regressions. (These tests also add indirect coverage for the more "basic" features of CUDAFuture).

I didn't know how/where to add tests for C++ ATen stuff, so instead I added these tests to the Python RPC suite, using the torch.futures.Future wrapper. (It made sense in my mind because RPC is the main user of CUDAFuture). I'll gladly accept pointers to better ways of doing this.

Differential Revision: [D27887191](https://our.internmc.facebook.com/intern/diff/D27887191/)

[ghstack-poisoned]

sab-regime reviewed (commented):
****
